syntax = "proto3";

package bank;

service BranchService {
    rpc Deposit (DepositRequest) returns (Response);
    rpc Withdraw (WithdrawRequest) returns (Response);
    rpc Propagate (PropagateRequest) returns (Response);
}

message DepositRequest {
    int32 customer_id = 1;
    int32 amount = 2;
    int32 logical_clock = 3;
}

message WithdrawRequest {
    int32 customer_id = 1;
    int32 amount = 2;
    int32 logical_clock = 3;
}

message PropagateRequest {
    int32 branch_id = 1;
    int32 customer_request_id = 2;
    int32 amount = 3;
    string action = 4;
    int32 logical_clock = 5;
}

message Response {
    bool success = 1;
    int32 logical_clock = 2;
}